driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -352,8 +374,38 @@ public boolean isClosed() {
     @Override
     public <T> T sendAndReceive(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext,
                                 final RequestContext requestContext, final OperationContext operationContext) {
-        CommandEventSender commandEventSender;
 
+        if (!Authenticator.shouldAuthenticate(authenticator, this.description)) {
+            return sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
+        }
+        Supplier<T> retryableOperation = () ->
+                sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
+        try {
+            return retryableOperation.get();
+        } catch (MongoCommandException e) {
+            if (triggersReauthentication(e)) {
+                authenticated.set(false);
+                authenticator.reauthenticate(this);
+                authenticated.set(true);
+                return retryableOperation.get();
+            }
+            throw e;
+        }
```

Comment:
This code can be simplified

```suggestion
        Supplier<T> sendAndReceiveInternal = () -> sendAndReceiveInternal(
                message, decoder, sessionContext, requestContext, operationContext);
        try {
            return sendAndReceiveInternal.get();
        } catch (MongoCommandException e) {
            if (triggersReauthentication(e) && Authenticator.shouldAuthenticate(authenticator, this.description)) {
                authenticated.set(false);
                authenticator.reauthenticate(this);
                authenticated.set(true);
                return sendAndReceiveInternal.get();
            }
            throw e;
        }
```